lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -838,72 +848,77 @@ void PatchInOutImportExport::processInOutMessage(CallInst &callInst) {
+void PatchInOutImportExport::visitCallInstrs(std::vector<Function *> calleeFuncs) {
+  for (auto callee : calleeFuncs) {
+    auto mangledName = callee->getName();
+
+    if (mangledName.startswith(lgcName::InputImportBuiltIn)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processInputImportBuiltIn(*callInst);
+        }
+      }
+    } else if (mangledName.startswith(lgcName::InputImportGeneric)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processInputImportNonBuiltIn(*callInst, false);
+        }
+      }
+    } else if (mangledName.startswith(lgcName::InputImportInterpolant)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processInputImportNonBuiltIn(*callInst, true);
+        }
+      }
+    } else if (mangledName.startswith(lgcName::OutputImportGeneric)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processOutputImport(*callInst, false);
+        }
+      }
+    } else if (mangledName.startswith(lgcName::OutputImportBuiltIn)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processOutputImport(*callInst, true);
+        }
+      }
+    } else if (mangledName.startswith(lgcName::OutputExportGeneric)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processOutputExportGeneric(*callInst);
+        }
+      }
+    } else if (mangledName.startswith(lgcName::OutputExportBuiltIn)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processOutputExportBuiltIn(*callInst);
+        }
+      }
+    } else if (mangledName.startswith(lgcName::OutputExportXfb)) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processOutputExportXfb(*callInst);
+        }
+      }
+    } else if (mangledName.startswith("llvm.amdgcn.s.sendmsg")) {
+      for (auto user : callee->users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          if (callInst->getParent()->getParent() == m_entryPoint)
+            processInOutMessage(*callInst);
+        }
+      }
```

Comment:
Add `else llvm_unreachable("Unexpected callee")`?

Reply:
Thanks for the really valuable comments! for this specific comment, I don't plan to add an unreachable check. When I collect the candidate callees, I just use very coarse-grained check "igc.input" and "igc.output", when visiting the callees' users, I use fine-grained check against each callee that need check, for other possible unintentionally added candidate callees, we just skip it. That is the idea behind the code. Although currently the coarse-grained and fine-grained check is equal in practice. If you still prefer an assert, please share your thought with me. I will check the failures in the cts and address all other review comments as suggested.

Reply:
That's fine, thanks for the explanation.